development/workflow.rst

```diff
@@ -0,0 +1,668 @@
+
+- Peers are a good proxy for maintainability.
+- It's useful for everyone to be familiar with other parts of the system.
+- Good practices can be spread; bad practices can be deprecated.
+
+All developers are expected to make time to perform reviews.
+The System Architect can intervene, however, if a developer is overburdened with review responsibility.
+
+.. _workflow-code-review-process:
+
+Code review discussion
+----------------------
+
+Code review discussion should happen on the GitHub pull request, with the reviewer giving a discussion summary and conclusive thumbs-up on the JIRA ticket.
+
+GitHub pull requests are ideal venues for discussion since individual commit diffs can be annotated and referenced.
+Be sure to make comments only from the **Conversation** and **Files changed** tabs---*not the Commits tab*.
+Any comments on code patches from the Commits tab will be lost if the developer amends and force pushes commits to the pull request.
+
+Code reviews are a collaborative check-and-improve process.
+Reviewers do not hold absolute authority, nor can developers ignore the reviewer's suggestions.
+The aim is to discuss, iterate, and improve the pull request until the work is ready to be deployed on ``master``.
+
+If the review becomes stuck on a design decision, that aspect of the review can be elevated into an RFC to seek team-wide consensus.
+
+If an issue is outside the ticket's scope, the reviewer should file a new ticket.
+
+Once the iterative review process is complete, the reviewer should switch the JIRA ticket's state to **Reviewed**.
+If there are multiple reviewers, our convention is that each review removes her/his name from the Reviews list to indicate sign-off; the final reviewer switches the status to 'Reviewed.'
+This indicates the ticket is ready to be merged.
+
+Note that in many cases the reviewer will mark a ticket as **Reviewed** before seeing the requested changes implemented.
+This convention is used when the review comments are non-controversial; the developer can simply implement the necessary changes and self-merge.
+The reviewer does not need to be consulted for final approval in this case.
+
+.. _workflow-code-review-merge:
+
+Merging
+-------
+
+Putting a ticket in a **Reviewed** state gives the developer the go-ahead to merge the ticket branch.
+If it has not been done already, the developer should rebase the ticket branch against the latest master.
+During this rebase, we recommend squashing any fixup commits into the main commit implementing that feature.
+Git commit history should not record the iterative improvements from code review.
+If a rebase was required, a final check with Jenkins should be done.
+
+We **always use non-fast forward merges** so that the merge point is marked in Git history, with the merge commit containing the ticket number:
+
+.. code-block:: bash
+
+   git checkout master
+   git pull  # sanity check
+   git merge --no-ff tickets/DM-NNNN
+   git push
```

Comment:
Github doesn't tell you whether there have been any commits on master since the branch was made so you can't be certain you are going to end up with the proper testable history or not.